internal/metricdata/subscriber.go

```diff
@@ -0,0 +1,215 @@
+// SPDX-FileCopyrightText: 2022-present Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package metricdata
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"github.com/omec-project/metricfunc/internal/promclient"
+	"github.com/omec-project/metricfunc/logger"
+	"github.com/omec-project/metricfunc/pkg/metricinfo"
+)
+
+var (
+	smContextCount  uint64
+	smContextActive uint64
+)
+
+func incSMContextActive() uint64 {
+	atomic.AddUint64(&smContextActive, 1)
+	return smContextActive
+}
+
+func decSMContextActive() uint64 {
+	atomic.AddUint64(&smContextActive, ^uint64(0))
+	return smContextActive
+}
+
+func GetSMContextCount() uint64 {
+	atomic.AddUint64(&smContextCount, 1)
+	return smContextCount
+}
+
+func HandleSubscriberEvent(subsData *metricinfo.CoreSubscriberData, sourceNf metricinfo.NfType) {
+	switch subsData.Operation {
+	case metricinfo.SubsOpAdd:
+		storeSubscriber(&subsData.Subscriber, sourceNf)
+	case metricinfo.SubsOpMod:
+		updateSubscriber(&subsData.Subscriber, sourceNf)
+	case metricinfo.SubsOpDel:
+		deleteSubscriber(&subsData.Subscriber, sourceNf)
+	default:
+		logger.CacheLog.Errorf("unknown smf subsriber operation [%v] ", subsData.Operation)
+	}
+}
+
+func storeSubscriber(sub *metricinfo.CoreSubscriber, sourceNf metricinfo.NfType) error {
+	metricData.SubLock.Lock()
+	//defer metricData.SubLock.Unlock()
+	_, ok := metricData.Subscribers[sub.Imsi]
+	if !ok {
+		metricData.Subscribers[sub.Imsi] = sub
+
+		promclient.SetSmfSessStats(sub.SmfIp, sub.Slice, sub.Dnn, sub.UpfName, incSMContextActive())
+		logger.CacheLog.Debugf("storing subscriber with imsi [%s] ", sub.Imsi)
+		pushPrometheusCoreSubData(sub)
+		metricData.SubLock.Unlock()
+	} else {
+		metricData.SubLock.Unlock()
+		updateSubscriber(sub, sourceNf)
+	}
+
+	return nil
+}
+
+func updateSubscriber(sub *metricinfo.CoreSubscriber, sourceNf metricinfo.NfType) {
+	metricData.SubLock.Lock()
+	defer metricData.SubLock.Unlock()
+	if s, ok := metricData.Subscribers[sub.Imsi]; ok {
+
+		deletePrometheusCoreSubData(s)
+
+		if sourceNf == metricinfo.NfTypeSmf {
+			//SMF specific fields
+			fillSmfSubsriberData(sub, s)
+		} else if sourceNf == metricinfo.NfTypeAmf {
+			//AMF specific fields
+			fillAmfSubsriberData(sub, s)
+		}
+		pushPrometheusCoreSubData(s)
+	}
+}
+
+func deleteSubscriber(sub *metricinfo.CoreSubscriber, sourceNf metricinfo.NfType) error {
+	metricData.SubLock.Lock()
+	defer metricData.SubLock.Unlock()
+	imsi := sub.Imsi
+	s, ok := metricData.Subscribers[imsi]
+	if ok {
```

Comment:
Isn't it good. If you say
if !ok {
   return
}
and rest of the code remains simple flow..early return code is easy to maintain.

Reply:
Right